Mathlib/Algebra/Ring/SumsOfSquares.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2024 Florent Schaffhauser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Florent Schaffhauser
+-/
+
+import Mathlib.Algebra.Ring.Defs
+import Mathlib.Algebra.Group.Submonoid.Basic
+import Mathlib.Algebra.Order.Ring.Defs
+
+/-!
+# Sums of squares
+
+We introduce sums of squares in a type `R` endowed with an `[Add R]`, `[Zero R]` and `[Mul R]`
+instances. Sums of squares in `R` are defined by an inductive predicate `isSumSq : R → Prop`:
+`0 : R` is a sum of squares and if `S` is a sum of squares, then for all `a : R`, `a * a + S` is a
+sum of squares in `R`.
+
+## Main declaration
+
+- The predicate `isSumSq : R → Prop`, defining the property of being a sum of squares in `R`.
+
+## Auxiliary declarations
+
+- The type `SumSqIn R` : in an additive monoid with multiplication `R`, we introduce the type
+`SumSqIn R` as the submonoid of `R` whose carrier is the subset `{S : R | isSumSq S}`.
+
+## Theorems
+
+- `isSumSq.add`  : if `S1` and `S2` are sums of squares in `R`, then so is `S1 + S2`.
+- `SumSq.nonneg` : in a linearly ordered semiring `R` with an `[ExistsAddOfLE R]` instance, sums of
+squares are non-negative.
+
+-/
+
+/--
+In a type `R` with an addition, a zero element and a multiplication, the property of being a sum of
+squares is defined by an inductive predicate: `0 : R` is a sum of squares and if `S` is a sum of
+squares, then for all `a : R`, `a * a + S` is a sum of squares in `R`.
+-/
+
+@[mk_iff]
+inductive isSumSq {R : Type*} [Add R] [Zero R] [Mul R] : R → Prop
```

Comment:
```suggestion
-/
@[mk_iff]
inductive isSumSq {R : Type*} [Add R] [Zero R] [Mul R] (r : R) : Prop
```

Reply:
Hmm, this particular one won't work, I think.

Reply:
Ah sorry, you're right.